ocaml/idl/ocaml_backend/gen_api.ml

```diff
@@ -135,8 +163,47 @@ let add_set_enums types =
           if List.exists (fun ty2 -> ty2 = DT.Set ty) types then [ty] else [DT.Set ty; ty]
         | _ -> [ty]) types)
 
+let all_types_of highapi = DU.Types.of_objects (Dm_api.objects_of_api highapi)
+
+(** Returns a list of type sorted such that the first elements in the
+    list have nothing depending on them. Later elements in the list may
+    depend upon types earlier in the list *)
+let toposort_types highapi types =
+  let rec inner result remaining =
+    let rec references name = function
+      | DT.String
+      | DT.Int
+      | DT.Float
+      | DT.Bool
+      | DT.DateTime
+      | DT.Ref _
+      | DT.Enum _ -> false
+      | DT.Set ty -> references name ty
+      | DT.Map (ty, ty') -> (references name ty) || (references name ty')
+      | DT.Record record when record = name -> true
+      | DT.Record record ->
+        let all_fields = DU.fields_of_obj (Dm_api.get_obj_by_name highapi ~objname:record) in
+        List.exists (fun fld -> references name fld.DT.ty) all_fields
+      | DT.Option ty -> references name ty
+    in
+    let (ty_ref,ty_not_ref) =
+      List.partition (fun ty -> match ty with
+      | DT.Record name ->
+        let referencing = List.filter (references name) remaining in
+        List.length referencing > 1
+      | _ -> false) remaining
+    in
+    match ty_ref with
+    | [] -> result @ ty_not_ref
+    | _ -> inner (result @ ty_not_ref) ty_ref
+  in
+  let result = inner [] types in
+  assert(List.length result = List.length types);
+  assert(List.sort compare result = List.sort compare types);
+  result
+
```

Comment:
this looks familiar... and could do with quite careful review.

Reply:
is it copied from elsewhere in xapi-project?

Reply:
no: https://github.com/jonludlam/xen-api/commit/1cbe6ae058926bb740cc2fe7deddb9836af94011 - I didn't get around to making sure it was correct and submitting a PR - hence the comment about careful review :-)

Reply:
My understanding is that the function works like this (not necesarily efficiently):
* it checks whether each type references any of the remaining types
* the types that do not reference anything are added straight to the sorted list: it doesn't matter in what order we emit these as they do not reference eachother, they could reference types emitted earlier though
* the remaining types are recursively checked with the same algorithm, since the leaf types got removed each type will reference one less, and thus the recursion eventually terminates

If you got self-referencing types or other loops the function would fail with a stackoverflow, but I think thats acceptable since this still happens at "build" time (when generating the ocaml API).

Reply:
I got the same understanding. In the end we are relying on the fact that this is a build-time only thing and we can pay the little price it involves just that once. Also, it has not yet failed :D

Even without this, I think having circular references would have been a problem

Reply:
Yes, that's how it's supposed to work. However, when I wrote this I recall it being a bit of a 'trial-and-error' style of development - it took a while to get the resulting code to compile. And that's all the testing it had at the time; that the resulting code compiled. I'd quite like to see some testing around this code.

Reply:
You are right, if the code is wrong we would see a failure in compiling the generated code, so we know its not wrong at least for the types we currently have declared.

Reply:
Absolutely true. Perhaps that's sufficient (for now).